kernel: Improve cooperative and preemptive performance as per thread_metric benchmark

peter-mitsis opened the pull request:
This set of commits offers improvements to both the cooperative and preemptive results in the the thread_metric benchmark.

The following numbers were obtained with the thread_metric benchmark project on the disco_l475_iot1 board. MQ denotes use of CONFIG_SCHED_MULTIQ and DQ denotes use of CONFIG_SCHED_DUMB.

Before:
Time Period Total:  12436655  (Cooperative, MQ)
Time Period Total:  11268922  (Cooperative, DQ)
Time Period Total:  5730607   (Preemptive, MQ)
Time Period Total:  6404080   (Preemptive, DQ)

After:
Time Period Total:  16001868  (Cooperative, MQ) : +28.7 %
Time Period Total:  11595554  (Cooperative, DQ) : +2.9 %
Time Period Total:  6473166   (Preemptive, MQ) : +12.9 %
Time Period Total:  7018286   (Preemptive, DQ) : +9.6%

peter-mitsis committed:
kernel: Add routines for _THREAD_QUEUED bit

Adds routines for setting and clearing the _THREAD_QUEUED
thread_state bit.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Add method to dequeue from a dlist

Dequeuing from a doubly linked list is similar to removing an item
except that it does not re-initialize the dequeued node.

This comes in handy when sorting a doubly linked list (where the
node gets removed and re-added). In that circumstance, re-initializing
the node is required. Furthermore, the compiler does not always
'understand' this. Thus, when performance is critical, dequeuing
may be preferred to removing.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Clean up priority_q.h

Minor cleanups include ...
 1. Eliminating unnecessary if-defs and forward declarations
 2. Co-locating routines of the same queue type

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: inline z_sched_prio_cmp()

Inlines z_sched_prio_cmp() to get better performance.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Add custom scheduler yield routines

Adds customized yield implementations based upon the selected
scheduler (dumb, multiq or scalable). Although each follows the
same broad outline, some of them allow for additional tweaking
to extract maximal performance. For example, the multiq variant
improves the performance of k_yield() by about 20%.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: cache priority queue index on UP multiq

Even though calculating the priority queue index in the priority
multiq is quick, caching it allows us to extract an extra 2% in
terms of performance as measured by the thread_metric cooperative
benchmark.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Simplify clear_halting() on UP systems

There is no need for clear_halting() to do anything on UP systems.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: thread suspend/resume bail paths are unlikely

Gives a hint to the compiler that the bail-out paths in both
k_thread_suspend() and k_thread_resume() are unlikely events.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Inline halt_thread() and z_thread_halt()

Inlining these routines helps to improve the
performance of k_thread_suspend()

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

cfriedt commented on include/zephyr/sys/dlist.h:
```diff
@@ -492,6 +492,15 @@ static inline void sys_dlist_insert_at(sys_dlist_t *list, sys_dnode_t *node,
 	}
 }
 
+static inline void sys_dlist_dequeue(sys_dnode_t *node)
```
Isn't this equivalent to `sys_dlist_remove()`?

Why not simply remove the `init()` from here? It is

https://github.com/zephyrproject-rtos/zephyr/blob/049b243b51f354c27273094a798ad2c9f34481c1/include/zephyr/sys/dlist.h#L511

peter-mitsis replied:
The `init()` in the `remove()` case allows us to determine whether the node is part of any doubly linked list--see _sys_dnode_is_linked()_. However, there are cases where a node can be removed from one list only to be immediately inserted into a doubly linked list elsewhere. When such cases on a critical path, it may be desirable to have an option that does not re-initialize the node.

nashif replied:
maybe dequeue is the wrong name here, you are calling a generic api based on where you are using it (in priority queues), would be `sys_dlist_remove_no_init()` be more appropriate? :-) .

peter-mitsis replied:
This cleaned up branch version somehow lost that warning. I'll definitely add it back.

cfriedt replied:
Naming things is hard. As long as the documentation is clear

peter-mitsis commented on kernel/include/priority_q.h:
```diff
@@ -57,13 +49,108 @@ static ALWAYS_INLINE void z_priq_mq_remove(struct _priq_mq *pq, struct k_thread
+#ifndef CONFIG_SMP
+	sys_dnode_t *n = sys_dlist_peek_next_no_check(pq, &_current->base.qnode_dlist);
+
+	sys_dlist_dequeue(&_current->base.qnode_dlist);
```
Yes, but sys_dlist_dequeue() will not re-initialize the node. This is relevant here because re-initializing the node in this case results in unnecessary operations.

peter-mitsis commented on kernel/include/priority_q.h:
```diff
@@ -57,13 +49,108 @@ static ALWAYS_INLINE void z_priq_mq_remove(struct _priq_mq *pq, struct k_thread
+		n = sys_dlist_peek_next_no_check(pq, n);
+	}
+
+	sys_dlist_append(pq, &_current->base.qnode_dlist);
```
Since we know that the node is going to be re-added to queue, it is advantageous to use the dequeue method as it will not spend cycles re-initializing the node only for that re-initialization to be undone when the node is added back into the queue.

andyross replied:
FWIW, I'm a little surprised the compiler isn't able to figure that out on its own and elide the writes that will end up clobbered.  There might be an aliasing analysis being defeated that we could treat with `__restrict` and some refactoring to put the two list operations next to each other, instead of inventing a new API.

cfriedt replied:
I guess appending a node will always re-initialize the prev and next pointers. I just worried there might be some leakage here, but it doesn't seem to be the case.

andyross commented on kernel/sched.c:
```diff
@@ -158,8 +158,10 @@ static inline bool is_halting(struct k_thread *thread)
 /* Clear the halting bits (_THREAD_ABORTING and _THREAD_SUSPENDING) */
 static inline void clear_halting(struct k_thread *thread)
 {
+#if CONFIG_MP_MAX_NUM_CPUS > 1
```
Style: IMHO a regular `if()` would be cleaner and clearer here.

andyross replied:
Non-style: CONFIG_SMP is the better tunable to test here, not the number of cores.  Non-SMP multicore builds are a legal edge case, and you can only run Zephyr threads on one core, so it doesn't need this.

peter-mitsis replied:
As I understand it, CONFIG_SMP with but 1 core is also legit. Perhaps something like ...

`#if defined(CONFIG_SMP) && (CONFIG_MP_MAX_NUM_CPUS > 1)`

... or its `if ()` equalivent.

andyross commented on kernel/sched.c:
```diff
@@ -1265,39 +1265,44 @@ static void halt_thread(struct k_thread *thread, uint8_t new_state)
-			}
-			(void)z_abort_thread_timeout(thread);
-			unpend_all(&thread->join_queue);
+		if (likely(new_state == _THREAD_SUSPENDED)) {
```
How impactful is this patch to halt_thread()?  This code was excruciatingly hard to get right and coming at it with a refactoring hatchet just for a minor performance boost gives me the willies.  I don't see anything wrong, but...

I guess my strong preference would be to split this one patch out into a separate PR and give it a ton of testing in isolation.

peter-mitsis replied:
This one patch gave us a 9.7% performance boost in the thread_metric preemptive benchmark (multiq) on the disco_l475_iot1 board when compared to the numbers from the previous commit.

andyross commented on kernel/sched.c:
```diff
@@ -36,7 +36,7 @@ struct k_spinlock _sched_spinlock;
 
 static ALWAYS_INLINE void update_cache(int preempt_ok);
-static void halt_thread(struct k_thread *thread, uint8_t new_state);
+static ALWAYS_INLINE void halt_thread(struct k_thread *thread, uint8_t new_state);
```
If it's always going to be inlined, it shouldn't need a prototype (which, obviously, can't be inlined per the language spec, though in the post-LTO world many compilers are able to do so).  Take this out, and if it doesn't build let's fix the declaration order.

peter-mitsis replied:
In this case, both they are serving as forward declarations to work around the fact that each calls the other.

andyross commented on kernel/sched.c:
```diff
@@ -1258,7 +1258,7 @@ extern void thread_abort_hook(struct k_thread *thread);
  * @param new_state New thread state (_THREAD_DEAD or _THREAD_SUSPENDED)
  */
-static void halt_thread(struct k_thread *thread, uint8_t new_state)
+static ALWAYS_INLINE void halt_thread(struct k_thread *thread, uint8_t new_state)
```
I think I count three spots where this medium-sized function is called, so an ALWAYS_INLINE is going to have a non-trivial code size impact.  Did you look at how bad it is?  My gut says that this is too big to be inlining, but will defer to numbers.

peter-mitsis replied:
Inlining this code showed a 124 byte code size increase on disco_l475_iot1 and a 7.8% performance boost. On qemu_x86, the code size increased by 128 bytes.

andyross reviewed (changes requested):
Some notes.  The desire to see that halt_thread() refactor slow-walked and separated just barely rises to a -1 out of conservatism, just to get it a little more time to show stability impacts.

peter-mitsis commented on kernel/include/priority_q.h:
```diff
@@ -279,23 +302,21 @@ static ALWAYS_INLINE void z_priq_mq_yield(struct _priq_mq *pq)
 
-		sys_dlist_t *l = &pq->queues[i * NBITS + TRAILING_ZEROS(pq->bitmask[i])];
-		sys_dnode_t *n = sys_dlist_peek_head(l);
+	if (likely(index != 0xFFFFFFFF)) {
```
Good idea. The initial state though should probably be `K_NUM_THREAD_PRIO - 1`--the slot for the lowest priority thread so that the update works correctly.

This should also reduce the patch set by 1. :)

peter-mitsis commented on include/zephyr/kernel_structs.h:
```diff
@@ -122,6 +122,9 @@ struct _priq_rb {
 	sys_dlist_t queues[K_NUM_THREAD_PRIO];
 	unsigned long bitmask[PRIQ_BITMAP_SIZE];
+#ifndef CONFIG_SMP
+	unsigned int prio_index;
```
I'm currently thinking `cached_queue_index`

peter-mitsis commented:
The main changes in this updated patch set are ...
  1. Split out the halt_thread() reordering patch as per Andy's preference, so that it can be its own PR later.
  2. Updated the implementation for caching the queue index in the muti-queue.

peter-mitsis commented:
Apologies for all the updates. I kept realizing that I had forgotten a small detail to fix each time.  This should hopefully be it.

peter-mitsis commented:
Rebased to resolve merge conflicts.

cfriedt reviewed (approved):
I like the performance improvement. Hopefully my ack isn't too preemptive.

andyross reviewed (approved):
I don't think I have much to complain about left.  Let's get it in and start banging on it.
